File: src/main/java/kr/hailor/hailor/service/ReservationService.java
package kr.hailor.hailor.service;

import kr.hailor.hailor.dto.ReservationDto;
import kr.hailor.hailor.entity.*;
import kr.hailor.hailor.repository.DesignerRepository;
import kr.hailor.hailor.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.stereotype.Repository;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
@RequiredArgsConstructor
public class ReservationService {

    private final ReservationRepository reservationRepository;
    private final DesignerRepository designerRepository;
    private final UserRepository userRepository;

    // [필수] 예약 생성: 대면/비대면 옵션, 디자이너 선택, 예약 날짜/시간을 받아 예약을 생성함
    public Reservation createReservation(String userEmail, ReservationDto dto) {
        // 1. 예약 신청한 사용자(회원) 조회
        User user = userRepository.findByEmail(userEmail)
                .orElseThrow(() -> new RuntimeException("유저를 찾을 수 없습니다: " + userEmail));

        // 2. 선택한 디자이너 조회
        Designer designer = designerRepository.findById(dto.getDesignerId())
                .orElseThrow(() -> new RuntimeException("디자이너를 찾을 수 없습니다: ID=" + dto.getDesignerId()));

        // 3. 대면/비대면 옵션 파싱
        ConsultationType type;
        try {
            type = ConsultationType.valueOf(dto.getConsultationType().toUpperCase());
        } catch (IllegalArgumentException e) {
            throw new RuntimeException("유효하지 않은 컨설팅 타입: " + dto.getConsultationType());
        }

        // 4. 예약 금액 결정 (디자이너의 offlinePrice 또는 onlinePrice 선택)
        int price = (type == ConsultationType.OFFLINE)
                ? designer.getOfflinePrice()
                : designer.getOnlinePrice();

        // 5. (필요 시) 예약 중복 체크 등 추가 로직 구현 가능

        // 6. 예약 엔티티 생성 및 저장
        Reservation reservation = Reservation.builder()
                .user(user)
                .designer(designer)
                .type(type)
                .reservationDateTime(dto.getReservationDateTime())
                .price(price)
                .canceled(false)
                .build();

        return reservationRepository.save(reservation);
    }

    // ----------------------------------------------------------------
    // 아래의 메서드들은 예약 조회/상세/취소 기능으로 현재 작업 범위에 포함되지 않으므로,
    // 추후 개발자가 구현할 때 참고하거나 제거해도 무방합니다.
    // ----------------------------------------------------------------

    // 유저 예약 목록 조회
    public List<Reservation> getReservationsByUser(String UserEmail) {
        return reservationRepository.findByUser_Email(UserEmail);
    }

    // 예약 상세 조회
    @Repository
    public interface ReservationRepository extends JpaRepository<Reservation, Long> {
        Optional<Reservation> findByIdAndUser_Email(Long reservationId, String userEmail);

        List<Reservation> findByUser_Email(String userEmail);
    }

    // 예약 취소
    public void cancelReservation(Long reservationId, String userEmail) {
        Reservation reservation = reservationRepository.findByIdAndUser_Email(reservationId, userEmail)
                .orElseThrow(() -> new RuntimeException("예약이 존재하지 않거나 권한이 없습니다."));

        reservation.setStatus(ReservationStatus.valueOf("CANCELLED")); // 상태 값 변경
        reservationRepository.save(reservation);
    }
}

File: src/main/java/kr/hailor/hailor/service/ReservationService.kt
package kr.hailor.hailor.service

import kr.hailor.hailor.client.KakaoPayClient
import kr.hailor.hailor.client.KakaoPayStatus
import kr.hailor.hailor.controller.forAdmin.reservation.AdminReservationListResponse
import kr.hailor.hailor.controller.forAdmin.reservation.AdminReservationSearchRequest
import kr.hailor.hailor.controller.forUser.reservation.ReservationCancelRequest
import kr.hailor.hailor.controller.forUser.reservation.ReservationCreateRequest
import kr.hailor.hailor.controller.forUser.reservation.ReservationInfoDto
import kr.hailor.hailor.controller.forUser.reservation.UserReservationsResponse
import kr.hailor.hailor.controller.forUser.reservation.UserReservationsSearchRequest
import kr.hailor.hailor.enity.MeetingType
import kr.hailor.hailor.enity.PaymentMethod
import kr.hailor.hailor.enity.Reservation
import kr.hailor.hailor.enity.ReservationStatus
import kr.hailor.hailor.enity.User
import kr.hailor.hailor.exception.AlreadyDoneException
import kr.hailor.hailor.exception.AlreadyReservedDesignerSlotException
import kr.hailor.hailor.exception.DesignerNotFoundException
import kr.hailor.hailor.exception.InvalidMeetingTypeException
import kr.hailor.hailor.exception.InvalidReservationDateException
import kr.hailor.hailor.exception.NeedGoogleAccessTokenException
import kr.hailor.hailor.exception.ReservationNotFoundException
import kr.hailor.hailor.exception.TemporallyUnavailableException
import kr.hailor.hailor.repository.DesignerRepository
import kr.hailor.hailor.repository.ObjectStorageRepository
import kr.hailor.hailor.repository.ReservationRepository
import kr.hailor.hailor.util.GoogleMeetManager
import kr.hailor.hailor.util.LockUtil
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import java.util.concurrent.TimeUnit

@Service
class ReservationService(
    private val reservationRepository: ReservationRepository,
    private val designerRepository: DesignerRepository,
    private val lockUtil: LockUtil,
    private val kakaoPayClient: KakaoPayClient,
    private val objectStorageRepository: ObjectStorageRepository,
    private val googleMeetManager: GoogleMeetManager,
) {
    @Transactional
    fun createReservation(
        user: User,
        request: ReservationCreateRequest,
    ) {
        val designer = designerRepository.findById(request.designerId).orElseThrow { DesignerNotFoundException() }
        val lockKey = "reservation:${designer.id}:${request.reservationDate}"

        if ((designer.meetingType != MeetingType.OFFLINE_AND_ONLINE && designer.meetingType != request.meetingType) ||
            request.meetingType == MeetingType.OFFLINE_AND_ONLINE
        ) {
            throw InvalidMeetingTypeException()
        }
        if (request.reservationDate > LocalDate.now().plusDays(90) ||
            request.reservationDate.isBefore(LocalDate.now()) ||
            request.slot !in 0 until 20 ||
            (
                request.reservationDate.isEqual(LocalDate.now()) &&
                    LocalDateTime.now() >
                    LocalDateTime.of(
                        LocalDate.now(),
                        LocalTime.of(10 + request.slot / 2, if (request.slot % 2 == 0)0 else 30),
                    )
            )
        ) {
            throw InvalidReservationDateException()
        }
        if (!lockUtil.lock(
                lockName = lockKey,
                waitTime = 3,
                leaseTime = 3,
                timeUnit = TimeUnit.SECONDS,
            )
        ) {
            throw TemporallyUnavailableException()
        }
        if (reservationRepository.existsByDesignerIdAndReservationDateAndSlotAndStatusIn(
                request.designerId,
                request.reservationDate,
                request.slot,
                listOf(ReservationStatus.RESERVED, ReservationStatus.CONFIRMED, ReservationStatus.FINISHED),
            )
        ) {
            lockUtil.unlock(lockKey)
            throw AlreadyReservedDesignerSlotException()
        }
        reservationRepository.save(
            Reservation(
                user = user,
                designer = designer,
                reservationDate = request.reservationDate,
                slot = request.slot,
                meetingType = request.meetingType,
                status = ReservationStatus.RESERVED,
                paymentMethod = request.paymentMethod,
                price = if (request.meetingType == MeetingType.ONLINE) designer.onlinePrice else designer.offlinePrice,
            ),
        )
        lockUtil.unlock(lockKey)
    }

    @Transactional(readOnly = true)
    fun getReservations(
        user: User,
        request: UserReservationsSearchRequest,
    ): UserReservationsResponse {
        val reservations =
            if (request.lastId == null) {
                reservationRepository.findAllByUserOrderByIdDesc(user, Pageable.ofSize(request.size))
            } else {
                reservationRepository.findAllByIdLessThanAndUserOrderByIdDesc(request.lastId, user, Pageable.ofSize(request.size))
            }
        return UserReservationsResponse(
            reservations = reservationsToUserReservationsResponse(reservations.content),
        )
    }

    fun getRecentFinishedReservation(
        user: User,
        request: UserReservationsSearchRequest,
    ): UserReservationsResponse {
        val reservations = reservationRepository.findAllRecentFinishedReservation(user, request, LocalDateTime.now())
        return UserReservationsResponse(
            reservations = reservationsToUserReservationsResponse(reservations.content),
        )
    }

    private fun reservationsToUserReservationsResponse(reservations: List<Reservation>) =
        reservations.map {
            ReservationInfoDto(
                id = it.id,
                date = it.reservationDate,
                status = it.status,
                slot = it.slot,
                meetingType = it.meetingType,
                paymentMethod = it.paymentMethod,
                googleMeetLink = it.googleMeetLink,
                price = it.price,
                designer =
                    ReservationInfoDto.ReservationDesignerInfoDto.of(
                        it.designer,
                        objectStorageRepository.getDownloadUrl(it.designer.profileImageName),
                    ),
            )
        }

    fun getAdminReservations(request: AdminReservationSearchRequest): AdminReservationListResponse {
        val reservations =
            if (request.lastId == null) {
                reservationRepository.findAllByOrderByIdDesc(Pageable.ofSize(request.size))
            } else {
                reservationRepository.findAllByIdLessThanOrderByIdDesc(request.lastId, Pageable.ofSize(request.size))
            }
        return AdminReservationListResponse(
            reservations =
                reservations.content.map {
                    AdminReservationListResponse.ReservationInfoDto(
                        id = it.id,
                        date = it.reservationDate,
                        slot = it.slot,
                        status = it.status,
                        meetingType = it.meetingType,
                        paymentMethod = it.paymentMethod,
                        googleMeetLink = it.googleMeetLink,
                        price = it.price,
                        designer =
                            AdminReservationListResponse.ReservationInfoDto.ReservationDesignerInfoDto.of(
                                it.designer,
                                objectStorageRepository.getDownloadUrl(it.designer.profileImageName),
                            ),
                        user = AdminReservationListResponse.ReservationInfoDto.ReservationUserInfoDto.of(it.user),
                    )
                },
        )
    }

    @Transactional
    fun cancelReservation(
        user: User,
        id: Long,
        request: ReservationCancelRequest,
    ) {
        val reservation = reservationRepository.findById(id).orElseThrow { ReservationNotFoundException() }
        if (reservation.user.id != user.id) {
            throw ReservationNotFoundException()
        }
        if (reservation.status in
            setOf(ReservationStatus.CANCELED, ReservationStatus.FINISHED, ReservationStatus.REFUNDED, ReservationStatus.NEED_REFUND)
        ) {
            throw AlreadyDoneException()
        }
        if (reservation.status == ReservationStatus.RESERVED) {
            reservation.status = ReservationStatus.CANCELED
        } else if (reservation.status == ReservationStatus.CONFIRMED) {
            reservation.status = ReservationStatus.NEED_REFUND
        }
        if (reservation.googleCalendarEventId != null && reservation.googleMeetLink != null) {
            if (request.googleAccessToken == null) {
                throw NeedGoogleAccessTokenException()
            }
            googleMeetManager.deleteGoogleMeet(reservation.googleCalendarEventId!!, request.googleAccessToken)
        }
    }

    @Transactional
    fun refundReservation(id: Long) {
        val reservation = reservationRepository.findById(id).orElseThrow { ReservationNotFoundException() }
        if (reservation.status != ReservationStatus.NEED_REFUND) {
            throw AlreadyDoneException()
        }
        reservation.status = ReservationStatus.REFUNDED
        if (reservation.paymentMethod == PaymentMethod.KAKAO_PAY) {
            val result = kakaoPayClient.getOrderStatus(reservation.paymentId!!)
            if (result.status == KakaoPayStatus.SUCCESS_PAYMENT) { // 결제가 완료된 경우에만 환불 진행
                kakaoPayClient.cancel(reservation.paymentId!!, reservation.price)
            }
        }
    }
}
